sklearn/cluster/_k_means_lloyd.pyx

```diff
@@ -0,0 +1,388 @@
+# cython: profile=True, boundscheck=False, wraparound=False, cdivision=True, initializedcheck=False
+# cython: language_level=3
+#
+# Licence: BSD 3 clause
+
+import numpy as np
+cimport numpy as np
+cimport openmp
+from cython cimport floating
+from cython.parallel import prange, parallel
+from libc.math cimport sqrt
+from libc.stdlib cimport malloc, free
+from libc.string cimport memset, memcpy
+from libc.float cimport DBL_MAX, FLT_MAX
+
+from ..utils.extmath import row_norms
+from ..utils._cython_blas cimport _gemm
+from ..utils._cython_blas cimport RowMajor, Trans, NoTrans
+from ._k_means cimport _relocate_empty_clusters_dense
+from ._k_means cimport _relocate_empty_clusters_sparse
+from ._k_means cimport _average_centers, _center_shift
+
+
+np.import_array()
+
+
+cpdef void _lloyd_iter_chunked_dense(np.ndarray[floating, ndim=2, mode='c'] X,
```

Comment:
Since this is C-like code I find it helpful to indicate whether the arguments are IN, OUT or IN OUT,

eg:
`floating[:, ::1] centers_old,  # OUT`

Kinda related: you can also use `const` in function definition and when cdefing a variable like `n_samples_chunk`

Reply:
@jeremiedbb can you please address this (here and elsewhere)